Ch3: wrangling

lheagy opened the pull request:
Another pass at the wrangling chapter. I have summarized comments / questions in #39. This is still going to need some work. I would appreciate input on the comments I made in #39. Feel free to push directly to this branch too! 

This PR:
- closes #39 
- closes #24

lheagy committed:
wip on ch3

lheagy committed:
working on wrangling chapter

lheagy committed:
move chaining content to the intro

lheagy committed:
Merge branch 'main' into ch3

lheagy committed:
update content on summary statistics

lheagy committed:
update the disucssion on apply

lheagy committed:
remove the discussion on What is a List

lheagy committed:
move the assign content to the very end

lheagy commented:
I took another crack at this! I moved the discussion of chaining to the intro chapter and have re-organized so that `assign` is at the very end. We are not worried about always ending up with a data frame at the end of each operation, so this simplifies the code. It is a pretty substantial re-write, so another set of eyes on the flow would be very much appreciated!

trevorcampbell committed:
minor wordsmithing on welcome page

trevorcampbell committed:
edited learning objs in ch1 for new chaining

trevorcampbell commented on source/intro.md:
```diff
@@ -631,20 +632,70 @@ We have now answered our initial question by generating this table!
+each language. When you move on to more complicated analyses, this issue only
+gets worse. In contrast, a *visualization* would convey this information in a much
+more easily understood format.
 Visualizations are a great tool for summarizing information to help you
```
this text is no longer valid here. need a better transition to chaining and move this argument regarding viz to right before the viz section.

trevorcampbell commented on source/intro.md:
````diff
@@ -631,20 +632,70 @@ We have now answered our initial question by generating this table!
+```{index} chaining methods
+```
+
+In Python, we often have to call multiple methods in a sequence to process a data
````
could be a smoother transition here and flow better from previous text. Can reference the fact that we need three ops (loc, sort, head), so what is the right way to do these?

trevorcampbell commented on source/intro.md:
```diff
@@ -631,20 +632,70 @@ We have now answered our initial question by generating this table!
+3) Sort the values by `mother_tongue` with the largest values at the top
+4) Ask for only the top 10 values using `head`
+
+One way of performing these three steps is to just write
```
looks like four steps in the list currently, but this says 3

also we should order these steps in the same way they were presented in this section (not sure if currently order is the same)

lheagy replied:
:+1: will combine to 3. The order is the same if you start from "Using `loc[]` to filter rows and select columns"

trevorcampbell commented on source/intro.md:
```diff
@@ -631,20 +632,70 @@ We have now answered our initial question by generating this table!
+4) Ask for only the top 10 values using `head`
+
+One way of performing these three steps is to just write
+multiple lines of code, storing temporary objects as you go:
```
publisher asked us to avoid putting ":" before code blocks, if I recall. Can just end the sentence with a period

trevorcampbell commented on source/intro.md:
````diff
@@ -631,20 +632,70 @@ We have now answered our initial question by generating this table!
+```
+This requires creating multiple temporary, intermediate outputs, and can get quite long.
+
+An alternative approach uses "chaining" to link multiple operations together. This can
````
unnecessary paragraph break

trevorcampbell commented on source/intro.md:
````diff
@@ -631,20 +632,70 @@ We have now answered our initial question by generating this table!
+arranged_lang_sorted = aboriginal_lang.sort_values(by='mother_tongue', ascending=False)
+ten_lang = arranged_lang_sorted.head(10)
+```
+This requires creating multiple temporary, intermediate outputs, and can get quite long.
````
- hard to read ("quite long" is saying this only implicitly)
- suggests that the named objects are of some importance, when really they are just intermediates

trevorcampbell commented on source/intro.md:
```diff
@@ -631,20 +632,70 @@ We have now answered our initial question by generating this table!
+    can_lang.loc[can_lang["category"] == "Aboriginal languages", ["language", "mother_tongue"]]
+    .sort_values(by="mother_tongue", ascending=False)
+    .head(10)
+    )
```
This block of code needs way more explanation. Like really break it down. What do the dots mean? why can I keep using dots over and over again? Why do I put each on its own line? is it another expression? etc etc etc

trevorcampbell commented on source/wrangling.md:
```diff
@@ -55,57 +55,19 @@ By the end of the chapter, readers will be able to do the following:
       - `.str.split`
   - Recall and use the following operators for their
     intended data wrangling tasks:
```
check the learning objectives for validity.
- "define what lists, series, and data frames are" -- do we do lists explicitly now, e.g.?

lheagy replied:
updated the content, so now we explicitly do lists and dictionaries

trevorcampbell commented on source/wrangling.md:
```diff
@@ -55,57 +55,19 @@ By the end of the chapter, readers will be able to do the following:
 
 ## Data frames, series, and lists
 
 In Chapters {ref}`intro` and {ref}`reading`, *data frames* were the focus:
```
The code
```
In Chapter {ref}`blah`, we do...
```
does not typeset properly in jupyterbook. See #50 . Need to search through whole chapter and change to

```
the {ref}`blah` chapter
```

trevorcampbell commented on source/wrangling.md:
```diff
@@ -152,45 +114,29 @@ data set. There are 13 entities in the data set in total, corresponding to the
-
-+++ {"tags": ["remove-cell"]}
-
-Python stores the columns of a data frame as either
```
We need to keep some amount of this text (rewritten appropriately) to say that dataframes use series as columns, otherwise students won't know why you are covering series below

lheagy replied:
:+1: see the section on Data Structures in Python (starting at line 247)

trevorcampbell commented on source/wrangling.md:
```diff
@@ -152,45 +114,29 @@ data set. There are 13 entities in the data set in total, corresponding to the
-You can create a `pandas` series using the `pd.Series()` function.  For 
-example, to create the vector `region` as shown in
-{numref}`fig:02-series`, you can write:
+In Python, `pandas` **series** are lists. They are ordered and can be indexed.
```
"series are lists"

I kind of prefer the R wording here: "are objects that can contain one or more elements"

also series technically not a python list

maybe write the R phrasing and then say something like "(essentially, a list)"

trevorcampbell replied:
also reinforce here that series are what pandas uses for each column of a df

trevorcampbell commented on source/wrangling.md:
```diff
@@ -152,45 +114,29 @@ data set. There are 13 entities in the data set in total, corresponding to the
-example, to create the vector `region` as shown in
-{numref}`fig:02-series`, you can write:
+In Python, `pandas` **series** are lists. They are ordered and can be indexed.
+They are strictly 1-dimensional and can contain any data type
```
"1-dimensional" will not mean anything to a first year. less jargon

trevorcampbell commented on source/wrangling.md:
```diff
@@ -152,45 +114,29 @@ data set. There are 13 entities in the data set in total, corresponding to the
-{numref}`fig:02-series`, you can write:
+In Python, `pandas` **series** are lists. They are ordered and can be indexed.
+They are strictly 1-dimensional and can contain any data type
+(integers, strings, floats, etc), including a mix of them; Python
```
although it can include a mix, we should make sure to remark that it is generally good practice to store only one kind of data in a series (because all observations of one variable should be the same type)

trevorcampbell commented on source/wrangling.md:
```diff
@@ -152,45 +114,29 @@ data set. There are 13 entities in the data set in total, corresponding to the
+{numref}`tab:datatype-table`.
+You can create a `pandas` series using the
+`pd.Series()` function.  For example, to create the series `region` as shown
+in{numref}`fig:02-series`, you can write:
```
spacing error

trevorcampbell commented on source/wrangling.md:
```diff
@@ -265,76 +176,26 @@ image_read("img/data_frame_slides_cdn/data_frame_slides_cdn.007.jpeg") %>%
+It is important in Python to make sure you represent your data with the correct type.
+Many of the `pandas` functions we use in this book treat
 the various data types differently. You should use integers and float types
 (which both fall under the "numeric" umbrella type) to represent numbers and perform
```
"numeric" isn't a type in python -- this was missed from R translation

trevorcampbell commented on source/wrangling.md:
```diff
@@ -265,76 +176,26 @@ image_read("img/data_frame_slides_cdn/data_frame_slides_cdn.007.jpeg") %>%
 arithmetic. Strings are used to represent data that should
-be thought of as "text", such as words, names, paths, URLs, and more. 
+be thought of as "text", such as words, names, paths, URLs, and more.
 There are other basic data types in Python, such as *set*
```
We need to discuss each type that appears in the table.

Missing bool, nonetype: what are these used for?

In the text here we should also reference the type name, so `int` `float` `bool` `str` and `NoneType` after we mention them in the sentence

E.g. 

"Strings (`str`) blahblah"

trevorcampbell commented on source/wrangling.md:
```diff
@@ -265,76 +176,26 @@ image_read("img/data_frame_slides_cdn/data_frame_slides_cdn.007.jpeg") %>%
+| integer               | `int`        | positive/negative whole numbers               | `42`                                       |
+| floating point number | `float`      | real number in decimal form                   | `3.14159`                                  |
+| boolean               | `bool`       | true or false                                 | `True`                                     |
+| string                | `str`        | text                                          | `"Can I have a cheezburger?"`              |
```
"Can I have a cheezburger" is cute but kind of out-of-place informality here

trevorcampbell commented on source/wrangling.md:
```diff
@@ -345,10 +206,10 @@ A vector versus a list.
 
 A data frame is really just series stuck together that follows two rules:
```
"is really just series stuck together"

series is its own plural here, so it's a bit awk

maybe

"A data frame is really just a list of series....."

lheagy replied:
I think the concept of a "list of series" might be a bit less intuitive than "stuck together" I will try a less backward wording, but if this doesn't fit with you, I am happy to adopt your suggestion

"
A data frame is really just a collection of series that are stuck together; it follows two rules
"

trevorcampbell commented on source/wrangling.md:
```diff
@@ -345,10 +206,10 @@ A vector versus a list.
-Not all columns in a data frame need to be of the same type. 
+Not all columns in a data frame need to be of the same type.
 {numref}`fig:02-dataframe` shows a data frame where
 the columns are series of different types.
```
Should take some of the text from the R version to put here at the end. Like

"But each element in one column should usually be the same type, since the values for a single variable are usually all of the same type. So even though series let you put different types in them, most common to have just one per column blahblah"

trevorcampbell commented on source/wrangling.md:
```diff
@@ -386,49 +230,29 @@ Data frame and vector types.
-:tags: [remove-cell]
-
-# The functions from the `tidyverse` package that we use often give us a
-# special class of data frame called a *tibble*. Tibbles have some additional \index{tibble}
```
We should keep this whole paragraph, but reword it to be about `pandas DataFrames`

There will be no mention of a "base python dataframe" because that doesn't exist here. But it's still worth saying what the type is, discussing that dataframes provide a bunch of functions (like loc and [] and whatever).

Try to keep the R pedagogy as much as possible

lheagy replied:
See new section on `### Data structures in Python` (starting on 251)

trevorcampbell commented on source/wrangling.md:
```diff
@@ -386,49 +230,29 @@ Data frame and vector types.
 > For example we can check the class of the Canadian languages data set,
```
the code cell below this note is not inside the note. It should be inside the note (see the R version)

lheagy replied:
hmm, that doesn't easily work in juypterbook... I created #88

joelostblom replied:
I think it would be nice if we used the [jupyter book admonitions](https://jupyterbook.org/en/stable/content/content-blocks.html#notes-warnings-and-other-admonitions) ("note" in this case) instead of block quotes for these section. These can also contain code as demonstrated in the link.

trevorcampbell commented on source/wrangling.md:
````diff
@@ -386,49 +230,29 @@ Data frame and vector types.
 type(can_lang)
 ```
 
 Lists, Series and DataFrames are basic types of *data structure* in Python, which
````
The table link here is broken (it points elsewhere in the chapter...)

also, should we cover `list`, `dict`, `tuple` and `set` here? This would probably be the place to do it. Also say "while this course will focus on data frames in Pandas, python has some very useful data structures that you will often see"

Table:
Data structure |  Example |   Usage
list   | [1,"a", 4] | storing an ordered collection of elements. You can index it like `blah`.
dict | {a : 3, blah} | storing a mapping from key to value

etc etc

lheagy replied:
I took a pass at this. Rather than including examples in the table, I included code examples below. See if you think this works

trevorcampbell commented on source/wrangling.md:
```diff
@@ -491,18 +315,18 @@ below!
+Data is often stored this way
+because this format is sometimes more intuitive for human readability
 and understanding, and humans create data sets.
-In {numref}`fig:02-wide-to-long`, 
```
The caption on the figure that this reference is pointing to is overlapping the image itself somehow (Fig 17). No idea why, but figure shouldn't overlap its caption

lheagy replied:
I have no idea why. I started #87 . It is still readable, so I don't think critical to solve at the moment

trevorcampbell commented on source/wrangling.md:
```diff
@@ -518,13 +342,16 @@ greatly simplified once the data is tidied.
-{numref}`fig:02-wide-to-long`.
+{numref}`fig:02-wide-to-long`. Note that the number of entries in our data frame
+can change in this transformation. The "untidy" data has 5 rows and 3 columns for
+a total of 15 data, whereas the "tidy" data on the right has 15 rows and 2 columns
```
We don't usually refer to entries in a data frame in this class as "data" (even though that's technically correct). I would use "entries" here

trevorcampbell commented on source/wrangling.md:
```diff
@@ -541,41 +368,42 @@ Melting data from a wide to long data format.
-The `.melt` function combines columns, 
-and is usually used during tidying data 
-when we need to make the data frame longer and narrower. 
+We say that we "melt" (or "pivot") the wide table into a longer format.
```
I would remove this new "We say that" sentence -- especially because "pivot" is actually a different function in python (the R naming is so much better in this case...pivot_wider and pivot_longer. Python has "melt" and "pivot").

trevorcampbell commented on source/wrangling.md:
````diff
@@ -613,13 +441,11 @@ We use `.melt` to combine the Toronto, Montréal,
 
 ```{code-cell} ipython3
 lang_mother_tidy = lang_wide.melt(
     id_vars=["category", "language"],
````
This code block is different from the image above

Need to add a bold black comment for melissa to update the image (if the current code block is what we want to go with)

lheagy replied:
I think it might be best to stick with the simpler code, so I will add a note

lheagy replied:
Note added in 481

trevorcampbell commented on source/wrangling.md:
```diff
@@ -613,13 +441,11 @@ We use `.melt` to combine the Toronto, Montréal,
 Canadians report each language as their mother tongue for each metropolitan
-area. We specify `value_vars` to be all
-the columns between Toronto and Edmonton:
+area:
```
avoid colon before code block

trevorcampbell commented on source/wrangling.md:
````diff
@@ -754,7 +581,7 @@ lang_home_tidy
 
 ```{code-cell} ipython3
-lang_home_tidy.dtypes
+lang_home_tidy.info()
````
holy crap this is a useful function!

We should absolutely teach `.info()` earlier. How about in a new section right before "Tidy data" (i.e., at the end of the "Data frames, series, and lists" section)

trevorcampbell replied:
(and of course explain all the printouts you get with `.info()`

lheagy replied:
added lines 655 and following to explain this. It doesn't seem like it needs its own heading? but we can easily add a heading if you prefer -- no strong opinions on my end

joelostblom replied:
I don't think this needs its own heading but agree that we should introduce it early. One possible spot is when we talk about data frame column types and the `type` function. We mention the importance of having the same type in each every column there, but we do'n't show them how to see which data types are in a column, which `info` would remedy.

trevorcampbell commented on source/wrangling.md:
````diff
@@ -55,57 +55,19 @@ By the end of the chapter, readers will be able to do the following:
-```
+      - `.iloc[]`
 
 ## Data frames, series, and lists
````
Do we cover lists any more? rename this section?

trevorcampbell commented on source/wrangling.md:
````diff
@@ -754,7 +581,7 @@ lang_home_tidy
 ```
````
Just above this: `lang_home_tidy.columns = [....]` is not explained at all. What is going on there? Do we need to do that any time we pivot? 

(also if we're renaming columns, we taught students how to do that in Ch 1 -- we should follow that example for consistency)

trevorcampbell replied:
`reset_index` is also not explained at all

lheagy replied:
added the text below. See what you think!

trevorcampbell commented on source/wrangling.md:
```diff
@@ -795,28 +622,34 @@ lang_messy
-and the other will contain the counts of Canadians 
-that speak each language most at work for each region. 
+Next we'll use `.str.split` to split the `value` column into two columns.
+How it works is that it takes a single string and splits it into multiple values
```
awk wording

trevorcampbell commented on source/wrangling.md:
```diff
@@ -795,28 +622,34 @@ lang_messy
-that speak each language most at work for each region. 
+Next we'll use `.str.split` to split the `value` column into two columns.
+How it works is that it takes a single string and splits it into multiple values
+based on the character you tell it to split on. For example:
```
avoid colon before code block

trevorcampbell commented on source/wrangling.md:
````diff
@@ -843,7 +676,7 @@ tidy_lang
 ```
````
again, lots of unexplained stuff going on above. `rename` and `drop`, `concat`, etc.

If we can simplify this code, we should

If we cannot (or can only to a small extent), we need to explain everything that is going on here in far more detail

trevorcampbell replied:
It's often a good idea to "look at the output" of a partial operation, go "oh man there's something wrong now!" and then fix it. This is way less mystical for students, because they see individually why each step is happening.

lheagy replied:
I broke this up into a few steps starting at 750. See if you think that is an improvement

joelostblom replied:
An alternative to concat + rename is to assign directly to two columns in the dataframe:

```py
lang_messy_longer[["most_at_home", "most_at_work"]] = lang_messy_longer["value"].str.split("/", expand=True)
```

I think this is a bit easier to read myself (after showing them the output from the right hand expression once first), but if we want to make a point of showing concat because we think it is generally useful then I think the current way looks good.

lheagy replied:
:+1: I like your suggestion @joelostblom: please feel free to change

trevorcampbell commented on source/wrangling.md:
````diff
@@ -843,7 +676,7 @@ tidy_lang
 
 ```{code-cell} ipython3
-tidy_lang.dtypes
+tidy_lang.info()
````
Do we need to print info here? or is the table print above enough?

trevorcampbell reviewed (changes requested):
I reviewed Ch1, and Ch3 up to "Using [] to extract rows or columns" (not including)

I figure I suggested enough changes that I will hold off on reviewing the rest until the current batch is sorted.

lheagy committed:
update discussion on chaining, edits from Trevor

lheagy committed:
update discussion of split

lheagy committed:
remove unnecessary call to File: dir	Node: Top	This is the top of the INFO tree
  This (the Directory node) gives a menu of major topics.
  Typing "d" returns here, "q" exits, "?" lists all INFO commands, "h"
  gives a primer for first-timers, "mEmacs<Return>" visits the Emacs topic,
  etc.
  In Emacs, you can click mouse button 2 on a menu item or cross reference
  to select it.
  --- PLEASE ADD DOCUMENTATION TO THIS TREE. (See INFO topic first.) ---

* Menu: The list of major topics begins on the next line.

Emacs
* Ada mode: (ada-mode). The GNU Emacs mode for editing Ada.
* Autotype: (autotype). Convenient features for text that you enter frequently
                          in Emacs.
* CC Mode: (ccmode).   Emacs mode for editing C, C++, Objective-C,
                          Java, Pike, and IDL code.
* CL: (cl).		Partial Common Lisp support for Emacs Lisp.
* Dired-X: (dired-x).   Dired Extra Features.
* EUDC: (eudc).   A client for directory servers (LDAP, PH)
* Ebrowse: (ebrowse).   A C++ class browser for Emacs.
* Ediff: (ediff).       A visual interface for comparing and merging programs.
* Emacs: (emacs).	The extensible self-documenting text editor.
* Emacs FAQ: (efaq).	Frequently Asked Questions about Emacs.
* Emacs MIME: (emacs-mime).   The MIME de/composition library.
* Eshell: (eshell).     A command shell implemented in Emacs Lisp.
* Forms: (forms).	Emacs package for editing data bases
			  by filling in forms.
* Gnus: (gnus).         The newsreader Gnus.
* IDLWAVE: (idlwave).	Major mode and shell for IDL and WAVE/CL files.

* MH-E: (mh-e).		Emacs interface to the MH mail system.
* Message: (message).   Mail and news composition mode that goes with Gnus.
* PCL-CVS: (pcl-cvs).	Emacs front-end to CVS.
* RefTeX: (reftex).	Emacs support for LaTeX cross-references and citations.
* SC: (sc).		Supercite lets you cite parts of messages you're
			  replying to, in flexible ways.

* Speedbar: (speedbar).        File/Tag summarizing utility.
* VIP: (vip).		An older VI-emulation for Emacs.
* VIPER: (viper).       The newest Emacs VI-emulation mode.
                          (also, A VI Plan for Emacs Rescue
                           or the VI PERil.)
* Widget: (widget).      Documenting the "widget" package used by the
                           Emacs Custom facility.
* WoMan: (woman).       Browse UN*X Manual Pages `Wo (without) Man'.

Texinfo documentation system
* Info: (info).                 Documentation browsing system.

Miscellaneous
* Screen: (screen).             Full-screen window manager.
* Standards: (standards).        GNU coding standards.

GNU admin
* Autoconf: (autoconf).         Create source code configuration scripts

Individual utilities
* aclocal: (automake)Invoking aclocal.          Generating aclocal.m4
* autoconf: (autoconf)autoconf Invocation.
                                How to create configuration scripts
* autoreconf: (autoconf)autoreconf Invocation.
                                Remaking multiple `configure' scripts
* autoscan: (autoconf)autoscan Invocation.
                                Semi-automatic `configure.ac' writing
* config.status: (autoconf)config.status Invocation.
                                Recreating a configuration
* configure: (autoconf)configure Invocation.
                                Configuring a package
* ifnames: (autoconf)ifnames Invocation.
                                Listing the conditionals in source code

GNU programming tools
* automake: (automake).		Making Makefile.in's

Utilities
* Bash: (bash).                     The GNU Bourne-Again SHell.

GNU Packages
* Tar: (tar).			Making tape (or disk) archives.

Individual utilities
* tar: (tar)tar invocation.                     Invoking GNU `tar'

Software development
* Cpp: (cpp).		       The GNU C preprocessor.
* Cpplib: (cppinternals).      Cpplib internals.
* gcc: (gcc).                  The GNU Compiler Collection.
* gccinstall: (gccinstall).    Installing the GNU Compiler Collection.
* gccint: (gccint).            Internals of the GNU Compiler Collection.
* gfortran: (gfortran).                  The GNU Fortran Compiler.

GNU Libraries
* libgomp: (libgomp).                    GNU OpenMP runtime library

Programming & development tools
* gdbm_dump: gdbm_dump(gdbm).   Dump the GDBM database into a flat file.
* gdbm_load: gdbm_load(gdbm).   Load the database from a flat file.

Utilities

GNU libraries
* gmp: (gmp).                   GNU Multiple Precision Arithmetic Library.

Software libraries
* GnuTLS: (gnutls).		GNU Transport Layer Security Library.
* GnuTLS-Guile: (gnutls-guile).		GNU Transport Layer Security Library. Guile bindings.
* libidn2: (libidn2).	Internationalized domain names (IDNA2008/TR46) processing.
* libtasn1: (libtasn1). Library for Abstract Syntax Notation One (ASN.1).
* mpfr: (mpfr).                 Multiple Precision Floating-Point Reliable Library.

GNU Packages
* mpc: (mpc)Multiple Precision Complex Library.

Development
* fftw3: (fftw3).	FFTW User's Manual.

Individual utilities
* aclocal-invocation: (automake)aclocal Invocation.   Generating aclocal.m4.
* autoconf-invocation: (autoconf)autoconf Invocation.
                                How to create configuration scripts
* autoheader: (autoconf)autoheader Invocation.
                                How to create configuration templates
* autom4te: (autoconf)autom4te Invocation.
                                The Autoconf executables backbone
* automake-invocation: (automake)automake Invocation. Generating Makefile.in.
* autoreconf: (autoconf)autoreconf Invocation.
                                Remaking multiple ‘configure’ scripts
* autoscan: (autoconf)autoscan Invocation.
                                Semi-automatic ‘configure.ac’ writing
* autoupdate: (autoconf)autoupdate Invocation.
                                Automatic update of ‘configure.ac’
* config.status: (autoconf)config.status Invocation. Recreating configurations.
* configure: (autoconf)configure Invocation.    Configuring a package.
* ifnames: (autoconf)ifnames Invocation.        Listing conditionals in source.
* libtool-invocation: (libtool)Invoking libtool. Running the 'libtool' script.
* libtoolize: (libtool)Invoking libtoolize.      Adding libtool support.
* testsuite: (autoconf)testsuite Invocation.    Running an Autotest test suite.

Software development
* Autoconf: (autoconf).         Create source code configuration scripts.
* Automake: (automake).         Making GNU standards-compliant Makefiles.
* Automake-history: (automake-history). History of Automake development.
* GNU libtextstyle: (libtextstyle).     Output of styled text.
* GNU libunistring: (libunistring).     Unicode string library.
* Libtool: (libtool).           Generic shared library support script.

Localization
* idn2: (libidn2)Invoking idn2.	Internationalized Domain Name (IDNA2008/TR46) conversion.

Encryption
* Nettle: (nettle).             A low-level cryptographic library.

System Administration
* certtool: (gnutls)certtool Invocation.	Manipulate certificates and keys.
* gnutls-cli: (gnutls)gnutls-cli Invocation.	GnuTLS test client.
* gnutls-cli-debug: (gnutls)gnutls-cli-debug Invocation.	GnuTLS debug client.
* gnutls-serv: (gnutls)gnutls-serv Invocation.	GnuTLS test server.
* psktool: (gnutls)psktool Invocation.	Simple TLS-Pre-Shared-Keys manager.
* srptool: (gnutls)srptool Invocation.	Simple SRP password tool.

Libraries
* libgpg-error: (gnupg).   Error codes and common code for GnuPG.

GNU Libraries
* libgcrypt: (gcrypt).  Cryptographic function library.

C++ libraries
* autosprintf: (autosprintf).   Support for printf format strings in C++.

GNU Gettext Utilities
* ISO3166: (gettext)Country Codes.             ISO 3166 country codes.
* ISO639: (gettext)Language Codes.             ISO 639 language codes.
* autopoint: (gettext)autopoint Invocation.    Copy gettext infrastructure.
* envsubst: (gettext)envsubst Invocation.      Expand environment variables.
* gettext: (gettext).                          GNU gettext utilities.
* gettextize: (gettext)gettextize Invocation.  Prepare a package for gettext.
* msgattrib: (gettext)msgattrib Invocation.    Select part of a PO file.
* msgcat: (gettext)msgcat Invocation.          Combine several PO files.
* msgcmp: (gettext)msgcmp Invocation.          Compare a PO file and template.
* msgcomm: (gettext)msgcomm Invocation.        Match two PO files.
* msgconv: (gettext)msgconv Invocation.        Convert PO file to encoding.
* msgen: (gettext)msgen Invocation.            Create an English PO file.
* msgexec: (gettext)msgexec Invocation.        Process a PO file.
* msgfilter: (gettext)msgfilter Invocation.    Pipe a PO file through a filter.
* msgfmt: (gettext)msgfmt Invocation.          Make MO files out of PO files.
* msggrep: (gettext)msggrep Invocation.        Select part of a PO file.
* msginit: (gettext)msginit Invocation.        Create a fresh PO file.
* msgmerge: (gettext)msgmerge Invocation.      Update a PO file from template.
* msgunfmt: (gettext)msgunfmt Invocation.      Uncompile MO file into PO file.
* msguniq: (gettext)msguniq Invocation.        Unify duplicates for PO file.
* ngettext: (gettext)ngettext Invocation.      Translate a message with plural.
* xgettext: (gettext)xgettext Invocation.      Extract strings into a PO file.

The Algorithmic Language Scheme
* Guile Reference: (guile).     The Guile reference manual.
* R5RS: (r5rs).                 The Revised(5) Report on Scheme.

lheagy committed:
take care of colons preceding code blocks

lheagy committed:
take care of chapter references

lheagy committed:
add discussion of lists and dicts

lheagy committed:
add table and discussion on basic data structures in python:

lheagy committed:
add description of info

lheagy commented on source/intro.md:
````diff
@@ -619,32 +620,88 @@ arranged_lang
+    )
+```
+
+What just happened? The code above starts with parentheses `()` and is split over multiple lines.
````
added this paragraph here. See if you think this does the trick!

lheagy commented on source/wrangling.md:
````diff
@@ -2005,138 +1349,259 @@ An alternative to aggregating on a dataframe
+`.assign` is a useful function to modify or create new data frame columns.
+
+```{code-cell} ipython3
+lang_messy = pd.read_csv("data/region_lang_top5_cities_messy.csv")
+lang_messy_longer = lang_messy.melt(
+    id_vars=["category", "language"],
+    var_name="region",
+    value_name="value",
+)
+tidy_lang_obj = (
+    pd.concat(
+        (lang_messy_longer, lang_messy_longer["value"].str.split("/", expand=True)),
+        axis=1,
+    )
+    .rename(columns={0: "most_at_home", 1: "most_at_work"})
+    .drop(columns=["value"])
+)
````
I am not sure this is critical to repeat. We could just pick up with `official_langs`? 

But I can also appreciate that it might be useful to revisit and see it all together

lheagy committed:
some general cleanup in apply, assign

lheagy committed:
typo fix in the intro

lheagy commented on source/wrangling.md:
```diff
@@ -386,48 +242,74 @@ Data frame and vector types.
-are core to most data analyses. We summarize them in 
-{numref}`tab:datastructure-table`. There are several other data structures in the Python programming 
-language (*e.g.,* matrices), but these are beyond the scope of this book.
+### Data structures in Python
```
This may not need its own title, but I found it helpful. Feel free to remove if you think it is best to flow from the content about DataFrames above

joelostblom replied:
I think this title is helpful so I would vote for keeping it.

lheagy commented on source/wrangling.md:
```diff
@@ -265,76 +180,29 @@ image_read("img/data_frame_slides_cdn/data_frame_slides_cdn.007.jpeg") %>%
-(which both fall under the "numeric" umbrella type) to represent numbers and perform
-arithmetic. Strings are used to represent data that should
-be thought of as "text", such as words, names, paths, URLs, and more. 
+to represent numbers and perform arithmetic. Integers have no decimals while floats contain decimals.
+Booleans are variables that can only take on one of two values: True or False.
+Strings are used to represent data that should
+be thought of as "text", such as words, names, paths, URLs, and more.
+A `NoneType` is a special type in Python that is used to indicate no-value; this can occur when
+you have missing data.
```
added this per comment below

lheagy committed:
a couple of type fixes:

trevorcampbell committed:
polish chaining/multiline exps

trevorcampbell committed:
polish ch3 up to and incl tidy data

trevorcampbell committed:
polish indexing

trevorcampbell committed:
more polish ch3

trevorcampbell committed:
fix python exercises link

lheagy committed:
more on groupby

lheagy committed:
improve groupy and discussion of lambda functions

lheagy committed:
try re-ordering the assign and apply content

lheagy committed:
global find replace to remove . in naming conventions

trevorcampbell committed:
caption on fig24 fixed

trevorcampbell committed:
polish up to apply

lheagy committed:
cleanup through Using  to create new columns

joelostblom commented on source/wrangling.md:
```diff
@@ -152,45 +114,32 @@ data set. There are 13 entities in the data set in total, corresponding to the
 A data frame storing data regarding the population of various regions in Canada. In this example data frame, the row that corresponds to the observation for the city of Vancouver is colored yellow, and the column that corresponds to the population variable is colored blue.
```
It would be nice if this image also highlighted what a value is, for both R and Py

![image](https://user-images.githubusercontent.com/4560057/210589818-c349c17a-2c2f-41aa-ace8-32efc130f049.png)

joelostblom commented on source/wrangling.md:
```diff
@@ -386,46 +238,72 @@ Data frame and vector types.
+
+A `list` is an ordered collection of values. To create a list, we put the contents of the list in between
+square brackets `[]`, where each item of the list is separated by a comma. A `list` can contain values
+of different types. The example below contains 6 `str` entries.
```
```suggestion
of different types. The example below contains six `str` entries.
```

joelostblom commented on source/wrangling.md:
```diff
@@ -386,46 +238,72 @@ Data frame and vector types.
+### Data structures in Python
 
-+++
+The `Series` and `DataFrame` types are *data structure* in Python, which
```
```suggestion
The `Series` and `DataFrame` types are *data structures* in Python, which
```

joelostblom commented on source/wrangling.md:
```diff
@@ -437,15 +315,15 @@ language (*e.g.,* matrices), but these are beyond the scope of this book.
 
 There are many ways a tabular data set can be organized. This chapter will focus
 on introducing the **tidy data** format of organization and how to make your raw
```
```suggestion
There are many ways a tabular data set can be organized. The data frames we have looked at so far
have all been using the **tidy data** format. This chapter will focus on introducing this type of
data frame format of organization and walk through how to make your raw
```


I think the start of this tidy data section is repetition of the "What is a data frame" section where we already introduced the concept of rows, columns, and variables. Maybe we can keep only one of these illustrations and start this section off by referring to the previous one as in my suggestion above?

joelostblom commented on source/wrangling.md:
````diff
@@ -541,41 +424,41 @@ Melting data from a wide to long data format.
 ```
 
 We can achieve this effect in Python using the `.melt` function from the `pandas` package.
````
I think we decided to just use `method` instead of `.method`, so we should remove the prefixed `.` here and elsewhere

joelostblom commented on source/wrangling.md:
```diff
@@ -629,9 +512,10 @@ lang_mother_tidy
+> the next line. Each line ends with a comma `,` making it easier to read.
+> Splitting long lines like this across multiple lines is encouraged
 > as it helps significantly with code readability. Generally speaking, you should
 > limit each line of code to about 80 characters.
```
Aren't we already introducing method chaining in the intro section chapter? In that case we can remove this from here I think.

trevorcampbell replied:
I think it's fine (and often good) to repeat things like this -- but maybe we should link back to the explanation in Ch1?

lheagy committed:
through the summery

joelostblom commented on source/wrangling.md:
```diff
@@ -698,14 +582,14 @@ lang_long = pd.read_csv("data/region_lang_top5_cities_long.csv")
+Canadians reporting their primary language at home and work.
 Doing that would be difficult with this data in its current form,
 since these two variables are stored in the same column.
 {numref}`fig:img-pivot-wider-table` shows how this data
```
It would be nice if this figure showed the params on separate lines as with the code block below and the image for `melt`

joelostblom replied:
Also, it is unnecessary to specify `values` here since all remaining columns will be used for the values parameter. This is the same reason we omitted `value_vars` above, so I think it would make sense to omit it here too for consistency and simplicity. I think we can add a sentence mentioning that these parameter exists for both sections

joelostblom commented on source/wrangling.md:
```diff
@@ -753,8 +639,27 @@ lang_home_tidy.columns = [
+where each row is "named" with an integer. This is a subtle point, but the main take-away is that
+when you call `pivot`, it is a good idea to call `reset_index` afterwards.
+
+The second operation we applied is to rename the columns. When we perform the `pivot`
+operation, it keeps the original column name `"count"` and adds the `"type"` as a second column name.
+Having two names for a column can be confusing! So we rename giving each column only one name.
```
I think it won't be clear to students why we do this when we previously used `rename`. I don't think there is a better way to flatten an index, [most other approaches ](https://datascientyst.com/flatten-multiindex-in-pandas/)would take more explaining. But it might make sense to explicitly say what we are doing here since it is different from what we taught them previously

lheagy committed:
add :tags: [output_scroll] for large code outputs, change figure types

lheagy committed:
trim vertical whitespace on figures:

joelostblom commented on source/wrangling.md:
````diff
@@ -795,57 +701,88 @@ lang_messy
+```{code-cell} ipython3
+tidy_lang = pd.concat(
+        (lang_messy_longer, split_counts),
         axis=1,
````
```suggestion
tidy_lang = pd.concat(
    [lang_messy_longer, split_counts],
    axis=1,
```

Same indentation as elsewhere and change to a list because we don't teach tuples I think?

trevorcampbell replied:
agree!

joelostblom commented on source/wrangling.md:
```diff
@@ -854,175 +791,65 @@ Is this data set now tidy? If we recall the three criteria for tidy data:
 
 We can see that this data now satisfies all three criteria, making it easier to
 analyze. But we aren't done yet! Notice in the table, all of the variables are
```
```suggestion
analyze. But we aren't done yet! Although we can't see it in the data frame above, all of the variables are actually
```

The dtype is not shown in pandas as it is in dplyr

joelostblom commented on source/wrangling.md:
```diff
@@ -854,175 +791,65 @@ Is this data set now tidy? If we recall the three criteria for tidy data:
+`most_at_home` and `most_at_work` columns as a number (e.g., finding rows
+above a numeric threshold of a column).
+That won't be possible if the variable is stored as a `object`.
+Fortunately, the `pandas.to_numeric` function provides a natural way to fix problems
```
There is also `infer_objects` and `convert_dtypes` which should do this automatically (but for some reason doesn't work here it seems... might still be worth a mention)

joelostblom commented on source/wrangling.md:
````diff
@@ -1124,25 +951,6 @@ pd.Series(["Vancouver", "Toronto"]) == pd.Series(["Toronto", "Vancouver"])
-# >```
-```
-
 ### Extracting rows above or below a threshold using `>` and `<`
````
Should we also mention `>=` and <=`?

trevorcampbell replied:
yes

trevorcampbell replied:
but no need for a special section (or title) for it. Just mention it in this section

joelostblom commented on source/wrangling.md:
```diff
@@ -1152,1262 +960,890 @@ glue("census_popn", "{0:,.0f}".format(35151728))
 than French in Montréal according to the 2016 Canadian census.
 
-+++ {"tags": []}
+### Extracting rows using `query`
```
I wonder if it will be confusing that we are introducing a second way of doing the same thing... maybe we should add an explicit comment on that the other way is the default way? Or remove `query` unless we make heavy use of it later in the book

joelostblom commented on source/wrangling.md:
````diff
@@ -1152,1262 +960,890 @@ glue("census_popn", "{0:,.0f}".format(35151728))
 
-```{index} pandas.DataFrame; df[]
+(loc-iloc)=
+## Using `loc[]` to filter rows and select columns.
````
To me it seems like everything that is done is this section could be done with just `[]` and there is no need to involve `loc` here. We can save `loc` for only those situation when we are both filtering rows and selecting columns, similar to what we did in the intro chapter.

joelostblom commented on source/wrangling.md:
```diff
@@ -1152,1262 +960,890 @@ glue("census_popn", "{0:,.0f}".format(35151728))
-Vancouver, English was reported as the most common primary language used at
-home.  However, in Montréal, this does not seem to be the case!
--->
+## Aggregating data with `agg`
```
```suggestion
## Aggregating data
```

Since we cover more than just `agg` in this section

joelostblom commented on source/wrangling.md:
```diff
@@ -1152,1262 +960,890 @@ glue("census_popn", "{0:,.0f}".format(35151728))
+What if you want to calculate summary statistics on an entire dataframe? Well,
+it turns out that the functions in {numref}`tab:basic-summary-statistics`
+can be applied to a whole data frame!
+For example, we can ask for the number of rows that each column has using `count`.
```
Hmmm, is this a useful thing to show? Doesn't data frames require that all column have the same number of rows so we would use `info` or `shape` to get the number of rows in the data frame instead?

joelostblom commented on source/wrangling.md:
```diff
@@ -1152,1262 +960,890 @@ glue("census_popn", "{0:,.0f}".format(35151728))
+Not surprisingly, they are all the same. We could also ask for the `mean`, but
+some of the columns in `region_lang` contain string data with words like `"Vancouver"`
+and `"Halifax"`---for these columns there is no way for `pandas` to compute the mean.
+So we provide the keyword `numeric_only=True` so that it only computes the mean of columns with numeric values. This
```
How annoying that pandas is starting to require this, I quite liked the convenient of not having to specify it... it seems like a reasonable default to only take the average of numerical columns... I wonder what their rationale is.

joelostblom commented on source/wrangling.md:
```diff
@@ -1152,1262 +960,890 @@ glue("census_popn", "{0:,.0f}".format(35151728))
-to order the rows in the data frame by the values of one or more columns. 
-Here we pass the column name `most_at_home` to sort the data frame rows by the values in that column, in ascending order.
+Similarly, if there are only some columns that you would like to get summary statistics of,
+we can first use `loc[]` and then ask for the summary statistic. An example of this is illustrated in {numref}`fig:summarize-across`.
```
```suggestion
we can first use `[]` and then ask for the summary statistic. An example of this is illustrated in {numref}`fig:summarize-across`.
```

joelostblom commented on source/wrangling.md:
````diff
@@ -1152,1262 +960,890 @@ glue("census_popn", "{0:,.0f}".format(35151728))
+We use `loc[]` to specify the columns and then `agg` to ask for both the `mean` and `std`.
 ```{code-cell} ipython3
-:tags: [remove-cell]
+region_lang.loc[:, "mother_tongue":"lang_known"].agg(["mean", "std"])
````
```suggestion
region_lang["mother_tongue":"lang_known"].agg(["mean", "std"])
```

If we make the change I suggestion to this section above and only use loc for both filtering rows and selecting coluns at the same time

joelostblom commented on source/wrangling.md:
```diff
@@ -1152,1262 +960,890 @@ glue("census_popn", "{0:,.0f}".format(35151728))
 
+Lets say that we want to know
+the mean and standard deviation of all of the columns between `"mother_tongue"` and `"lang_known"`.
+We use `loc[]` to specify the columns and then `agg` to ask for both the `mean` and `std`.
```
```suggestion
We use `[]` to specify the columns and then `agg` to ask for both the `mean` and `std`.
```

joelostblom reviewed (changes requested):
This is looking great! I have provided a few suggestions, most of which are minor. I will go through the rest of the PR tomorrow.

lheagy commented on source/wrangling.md:
```diff
@@ -1152,1262 +974,807 @@ glue("census_popn", "{0:,.0f}".format(35151728))
-and region.
-
+This new data frame has the same `region` column as the `english_lang` data frame. The order of
+the cities is different, but that is okay, we can use the `merge` function in `pandas` to say
```
note that I added the introduction of `merge` here. It seems like a natural example but was not in the previous version, so please see what you think

trevorcampbell replied:
more wrangling more better!

lheagy committed:
Update source/wrangling.md

Co-authored-by: Joel Ostblom <joelostblom@users.noreply.github.com>

lheagy committed:
Apply suggestions from code review

Co-authored-by: Joel Ostblom <joelostblom@users.noreply.github.com>

trevorcampbell committed:
polishing ch3

trevorcampbell committed:
final polish on wrangling

trevorcampbell committed:
final polish on ch3 joel comments
